File: android/src/main/java/ee/forgr/capacitor_persistent_account/CapacitorPersistentAccountPlugin.java
package ee.forgr.capacitor_persistent_account;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.annotation.Permission;
import com.getcapacitor.annotation.PermissionCallback;

@CapacitorPlugin(name = "CapacitorPersistentAccount")
public class CapacitorPersistentAccountPlugin extends Plugin {

    private final String pluginVersion = "8.0.10";

    private CapacitorPersistentAccount implementation;

    @Override
    public void load() {
        super.load();
        implementation = new CapacitorPersistentAccount(getContext());
    }

    @PluginMethod
    public void readAccount(PluginCall call) {
        JSObject ret = new JSObject();
        try {
            CapacitorPersistentAccount.AccountData accountData = implementation.readAccount();
            if (accountData != null) {
                if (accountData.data != null) {
                    ret.put("data", new org.json.JSONObject(accountData.data.toString()));
                } else {
                    ret.put("data", org.json.JSONObject.NULL);
                }
                ret.put("accountName", accountData.accountName);
            } else {
                ret.put("data", org.json.JSONObject.NULL);
                ret.put("accountName", org.json.JSONObject.NULL);
            }
            call.resolve(ret);
        } catch (Exception e) {
            call.reject("Failed to read account: " + e.getMessage());
        }
    }

    @PluginMethod
    public void saveAccount(PluginCall call) {
        try {
            JSObject data = call.getObject("data");
            String accountName = call.getString("accountName");
            org.json.JSONObject json = data != null ? new org.json.JSONObject(data.toString()) : null;
            implementation.saveAccount(json, accountName);
            call.resolve();
        } catch (Exception e) {
            call.reject("Failed to save account: " + e.getMessage());
        }
    }

    @PluginMethod
    public void getPluginVersion(final PluginCall call) {
        try {
            final JSObject ret = new JSObject();
            ret.put("version", this.pluginVersion);
            call.resolve(ret);
        } catch (final Exception e) {
            call.reject("Could not get plugin version", e);
        }
    }
}

File: ios/Sources/CapacitorPersistentAccountPlugin/CapacitorPersistentAccountPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(CapacitorPersistentAccountPlugin)
public class CapacitorPersistentAccountPlugin: CAPPlugin, CAPBridgedPlugin {
    private let pluginVersion: String = "8.0.10"
    public let identifier = "CapacitorPersistentAccountPlugin"
    public let jsName = "CapacitorPersistentAccount"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "readAccount", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "saveAccount", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPluginVersion", returnType: CAPPluginReturnPromise)
    ]
    private let implementation = CapacitorPersistentAccount()

    @objc func readAccount(_ call: CAPPluginCall) {
        do {
            let result = try implementation.readAccount()
            call.resolve([
                "data": result.data ?? NSNull(),
                "accountName": result.accountName ?? NSNull()
            ])
        } catch {
            call.reject("Failed to read account: \(error.localizedDescription)")
        }
    }

    @objc func saveAccount(_ call: CAPPluginCall) {
        do {
            let accountName = call.getString("accountName")
            try implementation.saveAccount(call.getObject("data"), accountName: accountName)
            call.resolve()
        } catch {
            call.reject("Failed to save account: \(error.localizedDescription)")
        }
    }

    @objc func getPluginVersion(_ call: CAPPluginCall) {
        call.resolve(["version": self.pluginVersion])
    }

}
